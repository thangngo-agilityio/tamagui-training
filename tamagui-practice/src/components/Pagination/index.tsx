import { Button, Flex } from '@chakra-ui/react';

// Components
import { ArrowPaginationIcon } from '@/icons';

// Constants
import { DOTS, NEXT, PREV } from '@/constants';

interface PaginationProps {
  currentPage?: number;
  isDisabledPrev?: boolean;
  isDisableNext?: boolean;
  arrOfCurrButtons?: string[];
  onPageChange?: (direction: string) => void;
  onClickPage?: (currentPage: number) => void;
}

const PaginationComponent = ({
  currentPage = 1,
  arrOfCurrButtons = [],
  isDisabledPrev = false,
  isDisableNext = false,
  onPageChange = () => {},
  onClickPage = () => {},
}: PaginationProps) => {
  const handleNextPage = () => onPageChange(NEXT);

  const handlePrevPage = () => onPageChange(PREV);

  return (
    <Flex
      data-testid="pagination"
      justifyContent="space-between"
      mt={1}
      flexDir="row"
      flexWrap="wrap"
      gap={{ base: 3, default: 0 }}
    >
      <Flex
        alignItems="center"
        w={{
          base: arrOfCurrButtons.length > 1 ? '100%' : 'auto',
          '0.8sm': arrOfCurrButtons.length > 2 ? '100%' : 'auto',
          default: 'auto',
        }}
        gap="4px"
      >
        <Button
          width={{ base: '6px', default: '30px' }}
          height={{ base: '6px', default: '30px' }}
          data-testid="prev-button"
          aria-label="btn-prev"
          variant="iconSecondary"
          cursor={isDisabledPrev ? 'not-allowed' : ''}
          isDisabled={isDisabledPrev}
          onClick={handlePrevPage}
        >
          <ArrowPaginationIcon
            width="10px"
            height="10px"
            isExpanded={false}
            rotate="90deg"
          />
        </Button>
        <Flex
          alignItems="center"
          justifyContent="space-between"
          w={{ base: '100%', default: 'auto' }}
        >
          {arrOfCurrButtons.map((item: string, index: number) => {
            const isDots = item === DOTS;
            const isDisable = currentPage === +item || isDots;
            const hoverStyle = isDots
              ? {}
              : {
                  color: 'text.700',
                  bg: 'background.800',
                };
            const disableStyle = isDots
              ? {}
              : {
                  color: 'text.700',
                  bg: 'background.800',
                };
            const handleClickPage = () => onClickPage(+item);

            return (
              <Button
                key={`${item}-${index}`}
                data-testid={`page-${item}-button`}
                aria-label="btn-pages"
                isDisabled={isDisable}
                mx={0.5}
                p={{ base: '15px', lg: '25px' }}
                fontSize={{ base: 'sm', lg: 'md' }}
                bg={currentPage === +item ? 'background.800' : 'transparent'}
                color={currentPage === +item ? 'text.700' : 'text.200'}
                {...(isDots && { cursor: 'not-allowed' })}
                _hover={hoverStyle}
                _disabled={disableStyle}
                onClick={handleClickPage}
              >
                {item}
              </Button>
            );
          })}
        </Flex>
        <Button
          data-testid="next-button"
          aria-label="btn-next"
          variant="iconSecondary"
          cursor={isDisableNext ? 'not-allowed' : ''}
          isDisabled={isDisableNext}
          onClick={handleNextPage}
          width={{ base: '6px', default: '30px' }}
          height={{ base: '6px', default: '30px' }}
        >
          <ArrowPaginationIcon
            width="10px"
            height="10px"
            isExpanded={false}
            rotate="-90deg"
          />
        </Button>
      </Flex>
    </Flex>
  );
};

const Pagination = PaginationComponent;
export default Pagination;
