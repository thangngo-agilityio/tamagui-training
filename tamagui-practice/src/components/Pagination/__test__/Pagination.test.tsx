import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Components
import Pagination from '..';

const onPageChangeMock = jest.fn();
const onClickPageMock = jest.fn();

describe('Pagination render', () => {
  test('Should render match with snapshot.', () => {
    const { container } = render(
      <Pagination
        currentPage={1}
        isDisableNext={true}
        isDisabledPrev={true}
        arrOfCurrButtons={['1']}
        onPageChange={onPageChangeMock}
        onClickPage={onClickPageMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Handle click page', async () => {
    const { getByTestId } = render(
      <Pagination onPageChange={onPageChangeMock} />,
    );

    const nextPage = getByTestId('next-button');

    await userEvent.click(nextPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });

  it('Handle next page', async () => {
    const { getByTestId } = render(
      <Pagination onPageChange={onPageChangeMock} />,
    );

    const nextPage = getByTestId('next-button');

    await userEvent.click(nextPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });

  it('Handle prev page', async () => {
    const { getByTestId } = render(
      <Pagination currentPage={2} onPageChange={onPageChangeMock} />,
    );

    const nextPage = getByTestId('prev-button');

    await userEvent.click(nextPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });
});
