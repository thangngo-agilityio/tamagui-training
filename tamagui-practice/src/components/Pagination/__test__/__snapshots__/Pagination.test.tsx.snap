// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination render Should render match with snapshot. 1`] = `
<div>
  <div
    class="css-1oc7l50"
    data-testid="pagination"
  >
    <div
      class="css-iuzkhh"
    >
      <button
        aria-label="btn-prev"
        class="chakra-button css-11op8wz"
        data-testid="prev-button"
        disabled=""
        type="button"
      >
        <svg
          fill="none"
          height="10px"
          style="transform: rotate(90deg);"
          viewBox="0 0 15 9"
          width="10px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0.365273 0.522622C0.576242 0.311717 0.862339 0.193237 1.16065 0.193237C1.45896 0.193237 1.74505 0.311717 1.95602 0.522622L7.52477 6.09137L13.0935 0.522622C13.3057 0.317694 13.5899 0.2043 13.8849 0.206863C14.1798 0.209427 14.462 0.327742 14.6706 0.536326C14.8792 0.74491 14.9975 1.02707 15 1.32205C15.0026 1.61702 14.8892 1.90119 14.6843 2.11337L8.32015 8.4775C8.10918 8.6884 7.82308 8.80688 7.52477 8.80688C7.22646 8.80688 6.94037 8.6884 6.7294 8.4775L0.365273 2.11337C0.154369 1.9024 0.0358887 1.61631 0.0358887 1.318C0.0358887 1.01969 0.154369 0.733591 0.365273 0.522622Z"
            fill="gray"
          />
        </svg>
      </button>
      <div
        class="css-1fu46g5"
      >
        <button
          aria-label="btn-pages"
          class="chakra-button css-162oouz"
          data-testid="page-1-button"
          disabled=""
          type="button"
        >
          1
        </button>
      </div>
      <button
        aria-label="btn-next"
        class="chakra-button css-1egskbo"
        data-testid="next-button"
        disabled=""
        type="button"
      >
        <svg
          fill="none"
          height="10px"
          style="transform: rotate(-90deg);"
          viewBox="0 0 15 9"
          width="10px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0.365273 0.522622C0.576242 0.311717 0.862339 0.193237 1.16065 0.193237C1.45896 0.193237 1.74505 0.311717 1.95602 0.522622L7.52477 6.09137L13.0935 0.522622C13.3057 0.317694 13.5899 0.2043 13.8849 0.206863C14.1798 0.209427 14.462 0.327742 14.6706 0.536326C14.8792 0.74491 14.9975 1.02707 15 1.32205C15.0026 1.61702 14.8892 1.90119 14.6843 2.11337L8.32015 8.4775C8.10918 8.6884 7.82308 8.80688 7.52477 8.80688C7.22646 8.80688 6.94037 8.6884 6.7294 8.4775L0.365273 2.11337C0.154369 1.9024 0.0358887 1.61631 0.0358887 1.318C0.0358887 1.01969 0.154369 0.733591 0.365273 0.522622Z"
            fill="gray"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
